### PR TITLE
table: use pool to reuse temporal buffer store

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve? 
`UpdateRecord` will create a new `BufferStore` each time. The `WriteStmtBufs.BufStore` is too complicated and cannot reuse buffers between statement.

### What is changed and how it works?
Use a `sync.Pool` to pool these temporal buffers used by `UpdateRecord` and `AddRecord`.

### Check List 

Tests 
 - No code

Code changes

 - Has exported variable/fields change

### CODE REVIEW

**Comment by siddontang** [comment]
seem we need to take care of lightning @kennytm

File: table/tables/tables.go [hunk 1]
```diff
@@ -299,6 +299,22 @@ func (t *TableCommon) FirstKey() kv.Key {
 	return t.RecordKey(math.MinInt64)
 }
 
+var tempBufStorePool = sync.Pool{
```
Comment (qw4990) [inline table/tables/tables.go]: Should we use `util/localpool/LocalPool` instead of `sync.Pool` since elements in it will be cleared during golang GC.

**Comment by kennytm** [comment]
(at a glance it shouldn't affect Lightning very much. we can submit another PR if this incurs any performance regression in Lightning)

File: table/tables/tables.go (hunk 1 above)
Comment (bobotu) [reply table/tables/tables.go]: The objects will be removed from `sync.Pool` is not used for two rounds of GC. I think this is good to reduce memory footprint. As `LocalPool` cannot grow and shrink the pool size, I think `sync.Pool` is better for now.

**Comment by qw4990** [comment]
@bobotu Should we cherry-pick it to 4.0?

**Comment by bobotu** [comment]
@qw4990 We don't need this patch after https://github.com/pingcap/tidb/pull/15931 merged.

**Pull request closed without merging**